File: Difficulty/Medium/[39] Combination Sum/solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        self.dfs(res, candidates, target, [], 0)
        return res
        
    def dfs(self, res, candidates, target, combination, begin):
        if target < 0:
            return
        if target == 0:
            res.append(combination)
            return
        for i in range(begin, len(candidates)):
            self.dfs(res, candidates, target - candidates[i], combination+[candidates[i]], i)

File: Difficulty/Medium/[39] Combination Sum/solution.cpp
/*
 * @lc app=leetcode id=39 lang=cpp
 *
 * [39] Combination Sum
 */
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        sort(candidates.begin(), candidates.end());
        vector<int> combination;
        searchtree(res, candidates, target, combination, 0);
        return res;
    }
    
    void searchtree(vector<vector<int>>& res, vector<int>& candidates, int target, vector<int>& combination, int start){
        if (!target){
            res.push_back(combination);
            return;
        }
        for (int i = start; i < candidates.size() && target >= candidates[i]; ++i){
            combination.push_back(candidates[i]);
            searchtree(res, candidates, target - candidates[i], combination, i);
            combination.pop_back();
        }
    }
};